SDK: Cache the RTC foci from the well-known file.

pixlwave opened the pull request:
This PR makes the following changes:

- Refactor `ServerCapabilities` into `ServerInfo`
    - It had nothing to do with `/capabilities` and was confusing.
- Include the `/.well-known/matrix/client` response in the `ServerInfo` cache.
- Add `Client::rtc_foci` which uses the cached well-known.
- Add `Client::is_livekit_rtc_supported` to the FFI using the cache RTC foci.

Requires https://github.com/ruma/ruma/pull/2091 before merging

pixlwave commented on Cargo.toml:
Not intended to merge like this, it obviously needs to be updated once https://github.com/ruma/ruma/pull/2091 is available on the right branch.

bmarty commented:
I have got a crash testing this branch on Android:

> org.matrix.rustcomponents.sdk.InternalException: called `Option::unwrap()` on a `None` value

pixlwave commented:
> > org.matrix.rustcomponents.sdk.InternalException: called `Option::unwrap()` on a `None` value

Have you got a stack trace for this?

bmarty commented:
I have not a lot. Just this:

> log: thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value': /Users/bmarty/workspaces/matrix-rust-sdk/crates/matrix-sdk/src/client/mod.rs:1869

pixlwave commented:
Hmmm, well that line is unchanged in the PR and I'm not massively sure how this works.

https://github.com/matrix-org/matrix-rust-sdk/blob/e0f18e7c8e737f7747f6c59ce631cfbe91e309f0/crates/matrix-sdk/src/client/mod.rs#L1868-L1869

I set the new field I just added, maybe we can get some clarity in the review 🤔

bnjbvr commented:
> Hmmm, well that line is unchanged in the PR and I'm not massively sure how this works.
> 
> https://github.com/matrix-org/matrix-rust-sdk/blob/e0f18e7c8e737f7747f6c59ce631cfbe91e309f0/crates/matrix-sdk/src/client/mod.rs#L1868-L1869
> 
> I set the new field I just added, maybe we can get some clarity in the review 🤔

Let's break out the statement: `f(&guard).unwrap()`:

- `&guard` gives you information that's supposed to be accurate now
- `f(&guard)` will call the closure parameter on it, that returns a function. In the new case you've added, it's this function:

```rust
server_info
                .well_known
                .as_ref()
                .and_then(|well_known| well_known.rtc_foci.clone().into())
```

- if this callback function returns `None`, then this `.unwrap()` will fail.

I suppose that if the `.well-known` is `None` (or if `rtc_foci` can be `None`?), then this will return `None`, and cause the unwrap to fail.

pixlwave commented:
Ohhhh, I thought this line was about unlocking access to the `inner.caches` 😂🤦‍♂️

I see it clear as day now. Will rename some stuff.

Thanks Benji 🙏

bmarty commented:
With the latest commits, it's not crashing anymore, but `isLivekitRtcSupported()` is returning `false` when testing on a matrix.org account. I believe that the file https://matrix.org/.well-known/matrix/client has the expected field that should make the method return false. Am I missing something?

pixlwave commented:
Ah, this is presumably because the old value is still cached without a well-known. I've just seen the same thing. Does it work if you call `Client.resetServerInfo()` once?

bmarty commented:
> Does it work if you call `Client.resetServerInfo()` once?

It did not fix the issue, but I'll build the SDK again with the latest commits

pixlwave commented:
Huh, it's working great for me both on matrix.org and my test server which doesn't have any RTC foci 🤔

pixlwave commented:
Do you want to try `doug/rtc-focus-logs` and filter your logs for `[BENOIT]` and report back what you see?

bmarty reviewed (approved):
Working fine on Element X Android on a matrix.org account.

bmarty commented:
I have added [a commit](https://github.com/matrix-org/matrix-rust-sdk/pull/5167/commits/68f45b7cbca22a2e27b85450389e20961cc4eb83), please let me know if it's correct (I am not 💯 sure) and if we need anything else to be able to merge this PR. Thanks!

pixlwave commented:
I've rebased on `main` and squashed Benoit's change into the history so it doesn't get merged with my fork in the history.

@bmarty: I think the issue was running `cargo update` - I reverted the lock file and let RustRover load the project and the resulting lock file changes were much smaller 🤷‍♂️

pixlwave commented:
Changelogs updated (TIL you put them in reverse chronological order).

poljar commented on bindings/matrix-sdk-ffi/CHANGELOG.md:
```diff
@@ -10,6 +10,8 @@ All notable changes to this project will be documented in this file.
 
 Breaking changes:
 
+- `Client::reset_server_capabilities` has been renamed to `Client::reset_server_info`.
+  ([#5167](https://github.com/matrix-org/matrix-rust-sdk/pull/5167))
```
Well that's still in the wrong place, this PR is not part of the 0.12.0 release from two weeks ago.

pixlwave replied:
Ohhhh sorry I see what you're saying, I conflated the ordering comment with the "wrong place" comments… 🙈

poljar commented on crates/matrix-sdk-base/CHANGELOG.md:
```diff
@@ -37,6 +37,10 @@ No notable changes in this release.
 
 ### Refactor
 
+- The cached `ServerCapabilities` has been renamed to `ServerInfo` and
+  additionally contains the well-known response alongside the existing server versions.
+  Despite the old name, it does not contain the server capabilities.
+  ([#5167](https://github.com/matrix-org/matrix-rust-sdk/pull/5167))
```
This one is under the 0.11.0 release, which is happened two months ago.

poljar commented on crates/matrix-sdk/CHANGELOG.md:
```diff
@@ -65,6 +65,10 @@ All notable changes to this project will be documented in this file.
 
 ### Refactor
 
+- `ClientServerCapabilities` has been renamed to `ClientServerInfo`. Alongside this,
+  `Client::reset_server_info` is now `Client::reset_server_info` and `Client::fetch_server_capabilities`
+  is now `Client::fetch_server_versions`, returning the server versions response directly.
+  ([#5167](https://github.com/matrix-org/matrix-rust-sdk/pull/5167))
```
This one is under the 0.12.0 release.

pixlwave commented:
> No, we had plans to use `git-cliff` but ultimately it was too complicated of a setup.
> 
> I think towncrier would reintroduce some of the downsides `git-cliff` had.

What were the downsides of git-cliff? Towncrier doesn't use commits, just changelog entries as a text file so it mimicks what we're doing right now updating the CHANGELOG.md, but eradicates conflicts, accidentally rebasing entries into the wrong version and semi-automates the PR link generation (from the file number). It would be a *very* nice improvement to the workflow :)

We could quite easily set up a `changelog.d` directory per crate and then tweak the existing matrix-org [guide](https://github.com/matrix-org/sygnal/blob/main/CONTRIBUTING.md#changelog) for contributors to match.
